tests/settings/SettingsTest.cpp

```diff
@@ -0,0 +1,263 @@
+{
+    settings.add("test_setting", "42");
+    CPPUNIT_ASSERT_EQUAL(int(42), settings.get<int>("test_setting"));
+
+    settings.add("test_setting", "-1");
+    CPPUNIT_ASSERT_EQUAL(int(-1), settings.get<int>("test_setting"));
+}
+
+void SettingsTest::addSettingDoubleTest()
+{
+    settings.add("test_setting", "1234567.890");
+    CPPUNIT_ASSERT_EQUAL(double(1234567.89), settings.get<double>("test_setting"));
+}
+
+void SettingsTest::addSettingSizeTTest()
+{
+    settings.add("test_setting", "666");
+    CPPUNIT_ASSERT_EQUAL(size_t(666), settings.get<size_t>("test_setting"));
+}
+
+void SettingsTest::addSettingUnsignedIntTest()
+{
+    settings.add("test_setting", "69");
+    CPPUNIT_ASSERT_EQUAL(unsigned(69), settings.get<unsigned int>("test_setting"));
+}
+
+void SettingsTest::addSettingBoolTest()
+{
+    settings.add("test_setting", "true");
+    CPPUNIT_ASSERT_EQUAL(true, settings.get<bool>("test_setting"));
+
+    settings.add("test_setting", "on");
+    CPPUNIT_ASSERT_EQUAL(true, settings.get<bool>("test_setting"));
+
+    settings.add("test_setting", "yes");
+    CPPUNIT_ASSERT_EQUAL(true, settings.get<bool>("test_setting"));
+
+    settings.add("test_setting", "True");
+    CPPUNIT_ASSERT_EQUAL(true, settings.get<bool>("test_setting"));
+
+    settings.add("test_setting", "50");
+    CPPUNIT_ASSERT_EQUAL(true, settings.get<bool>("test_setting"));
+
+    settings.add("test_setting", "0");
+    CPPUNIT_ASSERT_EQUAL_MESSAGE("0 should cast to false.",
+                                 false, settings.get<bool>("test_setting"));
+
+    settings.add("test_setting", "false");
+    CPPUNIT_ASSERT_EQUAL(false, settings.get<bool>("test_setting"));
+
+    settings.add("test_setting", "False");
+    CPPUNIT_ASSERT_EQUAL(false, settings.get<bool>("test_setting"));
+}
+
+class Slice; //Forward declaration to save some time compiling.
+
+void SettingsTest::addSettingExtruderTrainTest()
+{
+    //Add a slice with some extruder trains.
+    std::shared_ptr<Slice> current_slice = std::make_shared<Slice>(0);
+    Application::getInstance().current_slice = current_slice.get();
+    current_slice->scene.extruders.emplace_back(0, nullptr);
+    current_slice->scene.extruders.emplace_back(1, nullptr);
+    current_slice->scene.extruders.emplace_back(2, nullptr);
+
+    settings.add("test_setting", "2");
+    CPPUNIT_ASSERT_EQUAL(&current_slice->scene.extruders[2], &settings.get<ExtruderTrain&>("test_setting"));
+
+    settings.add("extruder_nr", "1");
+    settings.add("test_setting", "-1"); //-1 should let it fall back to the current extruder_nr.
+    CPPUNIT_ASSERT_EQUAL_MESSAGE("If the extruder is negative, it uses the extruder_nr setting.",
+                                 &current_slice->scene.extruders[1], &settings.get<ExtruderTrain&>("test_setting"));
+}
+
+void SettingsTest::addSettingLayerIndexTest()
+{
+    settings.add("test_setting", "-4");
+    CPPUNIT_ASSERT_EQUAL_MESSAGE("LayerIndex settings start counting from 0, so subtract one.",
```

Comment:
This test is a bit weird. If we use frontend layer numbers, we shouldn't be testing the setting against a negative number as this should be impossible from a frontend perspective.

Also note that when a raft is enabled then the layer index setting doesnt correspond to the actual layer number in the layer view any more.

Reply:
I've split it up to test the negative layer indices separately from the subtracting one.

Reply:
OK. My original comment was a bit unclear, but whatever.

1) I notice that you use the message for the unit test as a place to write a comment. This message doesn't tell the tester what went wrong in case the test fails. Remember to write clearer messages in the future.

2) You don't have to test for negative numders, because the only layer index frontend setting has a minimum_value of 1. Moreover, the user is never exposed to negative layer numbers, so future LayerIndex settings should also be thusly limited.